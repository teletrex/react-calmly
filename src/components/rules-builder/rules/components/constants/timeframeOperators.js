/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../../constants';

const { NONE, RANGE, SINGLE_VALUE, WHEN_RELATION, WITHIN_PERIOD } = RULE_INPUT_TYPES;

export const TIMEFRAME_OPERATORS = [
  // Within:
  {
    inputType: WITHIN_PERIOD,
    label: 'within the last',
    value: 'WITHIN_THE_LAST',
  },

  // Today:
  {
    inputType: NONE,
    label: 'on today',
    value: 'ON_TODAY',
  },
  {
    inputType: NONE,
    label: 'before today',
    value: 'BEFORE_TODAY',
  },

  // Any time:
  {
    inputType: NONE,
    label: 'at any time',
    value: 'AT_ANY_TIME',
  },

  // Date:
  {
    inputType: SINGLE_VALUE,
    label: 'on',
    value: 'ON',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'after',
    value: 'AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'before',
    value: 'BEFORE',
  },

  // Date range:
  {
    inputType: RANGE,
    label: 'between',
    value: 'BETWEEN',
  },

  // When relation:
  {
    inputType: WHEN_RELATION,
    label: 'exactly',
    value: 'EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'more than',
    value: 'MORE_THAN',
  },
  {
    inputType: WHEN_RELATION,
    label: 'less than',
    value: 'LESS_THAN',
  },
];
