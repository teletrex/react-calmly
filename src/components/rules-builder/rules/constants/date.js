/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DATE_PATTERN = 'd{1,2}/d{1,2}/d{4}';
