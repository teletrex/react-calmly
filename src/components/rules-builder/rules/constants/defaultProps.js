/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const DEFAULT_PROPS = {
  INPUT_FIELD_VALIDATION: [],
  OPERATORS: [],
  TIMEFRAME: {},
};
