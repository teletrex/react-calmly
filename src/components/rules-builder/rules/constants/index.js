/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export * from './date';
export * from './defaultProps';
export * from './periods';
export * from './propTypes';
export * from './ruleOperators';
export * from './validationMessages';
export * from './values';
export * from './whenRelations';
