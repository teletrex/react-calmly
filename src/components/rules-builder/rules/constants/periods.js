/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




export const PERIODS = [
  {
    label: 'days',
    value: 'DAYS',
  },
  {
    label: 'weeks',
    value: 'WEEKS',
  },
  {
    label: 'months',
    value: 'MONTHS',
  },
];
