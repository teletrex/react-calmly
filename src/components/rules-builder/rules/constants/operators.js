/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import { RULE_INPUT_TYPES } from '../../constants';

const {
  MULTIPLE_VALUES,
  NONE,
  PERIOD_NUMBER,
  PERIOD_RANGE,
  RANGE,
  SINGLE_VALUE,
  WHEN_RELATION,
  WITHIN_PERIOD,
} = RULE_INPUT_TYPES;

// Common operators
export const BLANK_OPERATORS = [
  {
    inputType: NONE,
    label: 'is blank',
    value: 'IS_BLANK',
  },
  {
    inputType: NONE,
    label: 'is not blank',
    value: 'IS_NOT_BLANK',
  },
];

export const EXACT_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'is exactly',
    value: 'IS_EXACTLY',
  },
  {
    inputType: WHEN_RELATION,
    label: 'is not exactly',
    value: 'IS_NOT_EXACTLY',
  },
];

export const IS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is',
    value: 'IS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is not',
    value: 'IS_NOT',
  },
];

export const HAS_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'has',
    value: 'HAS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'has not',
    value: 'HAS_NOT',
  },
];

export const IS_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'is one of the following',
    value: 'IS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'is not one of the following',
    value: 'IS_NOT_ONE_OF_THE_FOLLOWING',
  },
];

export const CONTAIN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'contains',
    value: 'CONTAINS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not contain',
    value: 'DOES_NOT_CONTAIN',
  },
];

export const CONTAIN_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'contains one of the following',
    value: 'CONTAINS_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not contain one of the following',
    value: 'DOES_NOT_CONTAIN_ONE_OF_THE_FOLLOWING',
  },
];

export const RANGE_OPERATORS = [
  {
    inputType: RANGE,
    label: 'is between',
    value: 'IS_BETWEEN',
  },
];

// String related operators
export const STARTS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'starts with',
    value: 'STARTS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not start with',
    value: 'DOES_NOT_START_WITH',
  },
];

export const STARTS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'starts with one of the following',
    value: 'STARTS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not start with one of the following',
    value: 'DOES_NOT_START_WITH_ONE_OF_THE_FOLLOWING',
  },
];

export const ENDS_WITH_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'ends with',
    value: 'ENDS_WITH',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'does not end with',
    value: 'DOES_NOT_END_WITH',
  },
];

export const ENDS_WITH_ONE_OF_OPERATORS = [
  {
    inputType: MULTIPLE_VALUES,
    label: 'ends with one of the following',
    value: 'ENDS_WITH_ONE_OF_THE_FOLLOWING',
  },
  {
    inputType: MULTIPLE_VALUES,
    label: 'does not end with one of the following',
    value: 'DOES_NOT_END_WITH_ONE_OF_THE_FOLLOWING',
  },
];

// Number related operators
export const LESS_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is less than',
    value: 'IS_LESS',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is less than or equal to',
    value: 'IS_LESS_OR_EQUAL',
  },
];

export const GREATER_THAN_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than',
    value: 'IS_GREATER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is greater than or equal to',
    value: 'IS_GREATER_OR_EQUAL',
  },
];

// Date related operators
export const TODAY_OPERATORS = [
  {
    inputType: NONE,
    label: 'is today',
    value: 'IS_TODAY',
  },
  {
    inputType: NONE,
    label: 'is after today',
    value: 'IS_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not after today',
    value: 'IS_NOT_AFTER_TODAY',
  },
  {
    inputType: NONE,
    label: 'is before today',
    value: 'IS_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'is not before today',
    value: 'IS_NOT_BEFORE_TODAY',
  },
  {
    inputType: NONE,
    label: 'anniversary is today',
    value: 'ANNIVERSARY_IS_TODAY',
  },
];

export const ANNIVERSARY_OPERATORS = [
  {
    inputType: WHEN_RELATION,
    label: 'anniversary is',
    value: 'ANNIVERSARY_IS',
  },
];

export const WITHIN_OPERATORS = [
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the last',
    value: 'IS_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the last',
    value: 'IS_NOT_WITHIN_THE_LAST',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is within the next',
    value: 'IS_WITHIN_THE_NEXT',
  },
  {
    inputType: WITHIN_PERIOD,
    label: 'is not within the next',
    value: 'IS_NOT_WITHIN_THE_NEXT',
  },
];

export const WITHIN_RANGE_OPERATORS = [
  {
    inputType: PERIOD_RANGE,
    label: 'is within the range of the last',
    value: 'IS_WITHIN_THE_RANGE_OF_THE_LAST',
  },
];

export const PERIOD_EQUAL_OPERATORS = [
  {
    inputType: PERIOD_NUMBER,
    label: 'year is',
    value: 'YEAR_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'month is',
    value: 'MONTH_IS',
  },
  {
    inputType: PERIOD_NUMBER,
    label: 'day is',
    value: 'DAY_IS',
  },
];

export const BEFORE_AFTER_DATE_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'is after',
    value: 'IS_AFTER',
  },
  {
    inputType: SINGLE_VALUE,
    label: 'is before',
    value: 'IS_BEFORE',
  }
]

export const DATA_AGGREGATION_OPERATORS = [
  {
    inputType: SINGLE_VALUE,
    label: 'by',
    value: 'BY',
  },

];
