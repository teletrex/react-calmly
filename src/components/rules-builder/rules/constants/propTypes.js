/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import PropTypes from 'prop-types';

import { RULE_INPUT_TYPES } from '../../constants';

const { any, arrayOf, func, oneOf, oneOfType, shape, string } = PropTypes;

const OPERATOR = oneOfType([
  string,
  shape({
    inputType: oneOf(Object.values(RULE_INPUT_TYPES)).isRequired,
    label: string.isRequired,
    value: string.isRequired,
  }),
]);
const OPERATORS = arrayOf(OPERATOR);

const TIMEFRAME = shape({
  data: shape({
    operator: string,
    value: any,
  }),
  onChange: func,
});

const OPERATOR_NAME = string;

const SELECT_OPTION = shape({
  label: string.isRequired,
  value: string.isRequired,
});

const SELECT_OPTIONS = arrayOf(SELECT_OPTION);

export const PROPTYPES = {
  OPERATOR,
  OPERATORS,
  OPERATOR_NAME,
  SELECT_OPTION,
  SELECT_OPTIONS,
  TIMEFRAME,
};
