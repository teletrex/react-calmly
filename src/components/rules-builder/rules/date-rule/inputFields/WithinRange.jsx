/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React, { useContext, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import * as Yup from 'yup';
import isEqual from 'lodash/isEqual';

import { useTranslation } from '../../../../../translation';
import RulesBuilderContext from '../../../RulesBuilderContext';
import { NumberValue, PeriodDropdown } from '../../components/inputFields/components';
import { PERIODS, PROPTYPES, VALIDATION_MESSAGES } from '../../constants';
import { getPrefixedClasses, useFormValidation } from '../../helpers';

const MIN_VALUE1 = 0;
const MIN_VALUE2 = 1;

const WithinRange = ({ id, onValueChange, operator, value: { value, period: periodName } }) => {
  const INPUT_ID_1 = `${id}-1`;
  const INPUT_ID_2 = `${id}-2`;

  const RANGE_VALIDATION = {
    [INPUT_ID_1]: VALIDATION_MESSAGES.rangeFirst,
    [INPUT_ID_2]: VALIDATION_MESSAGES.rangeSecond,
  };

  const { t } = useTranslation();
  const { isReadOnly } = useContext(RulesBuilderContext);
  const [period, setPeriod] = useState(PERIODS.find(p => p.value === periodName));
  const [lastTouchedInput, setLastTouchedInput] = useState(null);

  const validateValues = fieldValues => {
    if (fieldValues[INPUT_ID_1] >= fieldValues[INPUT_ID_2]) {
      return { [lastTouchedInput]: t(RANGE_VALIDATION[lastTouchedInput]) };
    }

    return {};
  };

  const validationSchema = Yup.object().shape({
    [INPUT_ID_1]: Yup.number().min(MIN_VALUE1, VALIDATION_MESSAGES.min),
    [INPUT_ID_2]: Yup.number().min(MIN_VALUE2, VALIDATION_MESSAGES.min),
  });

  const {
    errors,
    handleBlur,
    initialValues,
    resetForm,
    setFieldValue,
    submitForm,
    values,
  } = useFormValidation({
    initialValues: {
      [INPUT_ID_1]: value[0],
      [INPUT_ID_2]: value[1],
    },
    validationSchema,
    validate: validateValues,
    onSubmit: values => {
      if (!isEqual(values, initialValues)) {
        onValueChange({
          value: [values[INPUT_ID_1], values[INPUT_ID_2]],
          period: period?.value,
        });
      }
    },
  });

  useEffect(() => {
    resetForm();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [operator]);

  const handlePeriodChange = newPeriod => {
    const { selectedItem } = newPeriod;
    setPeriod(selectedItem);

    onValueChange({
      value: [values[INPUT_ID_1], values[INPUT_ID_2]],
      period: selectedItem.value,
    });
  };

  const handleValueChange = ({ imaginaryTarget: { value: newValue, id: inputId } }) => {
    setLastTouchedInput(inputId);
    setFieldValue(inputId, +newValue);
  };

  const handleInputBlur = e => {
    handleBlur(e);
    submitForm();
  };

  if (isReadOnly) {
    const leftSpacingClass = getPrefixedClasses('--text-space-left');

    return (
      <span className={getPrefixedClasses('--rule-value-readonly')}>
        <span className={leftSpacingClass}>{values[INPUT_ID_1]}</span>
        <span className={leftSpacingClass}>to</span>
        <span className={leftSpacingClass}>{values[INPUT_ID_2]}</span>
        <span className={leftSpacingClass}>{period.label}</span>
      </span>
    );
  }

  return (
    <div
      className={getPrefixedClasses('--input-field-container')}
      data-testid="date-rule-range-of-last"
    >
      <NumberValue
        errors={errors}
        fieldName={INPUT_ID_1}
        id={INPUT_ID_1}
        onBlur={handleInputBlur}
        onChange={handleValueChange}
        value={value[0]}
      />
      <span
        className={getPrefixedClasses(['--text-space-left', '--text-space-right', '--rule-text'])}
      >
        to
      </span>

      <NumberValue
        errors={errors}
        fieldName={INPUT_ID_2}
        id={INPUT_ID_2}
        onBlur={handleInputBlur}
        onChange={handleValueChange}
        value={value[1]}
      />
      <PeriodDropdown idPrefix={id} onChange={handlePeriodChange} value={period} />
    </div>
  );
};

WithinRange.propTypes = {
  id: PropTypes.string.isRequired,
  onValueChange: PropTypes.func.isRequired,
  operator: PROPTYPES.OPERATOR,
  value: PropTypes.shape({
    period: PropTypes.string,
    value: PropTypes.arrayOf(PropTypes.number),
  }),
};

WithinRange.defaultProps = {
  operator: null,
  value: {
    period: PERIODS[0].value,
    value: [],
  },
};

export default WithinRange;
