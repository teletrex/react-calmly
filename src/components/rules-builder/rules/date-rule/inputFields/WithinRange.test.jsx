/* <LICENSE>
* 
* Copyright (C) 2022 Louis F. Roehrs, All rights reserved.
* 
* </LICENSE>
*  */




import React from 'react';
import { fireEvent, render, waitFor, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { PERIODS } from '../../constants';
import { renderWithContext } from '../../../utils/testHelpers';

import WithinRange from './WithinRange';

const initialPeriod = PERIODS[2];
const initialValue = {
  period: initialPeriod.value,
  value: [2, 7],
};

const props = {
  id: 'TestDateRuleElement',
  onValueChange: jest.fn(),
};

describe('WithinRange', () => {
  beforeEach(() => jest.resetAllMocks());

  it('should render correctly', () => {
    const { container } = render(<WithinRange {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly with a provided value', () => {
    const { container } = render(<WithinRange {...props} value={initialValue} />);
    expect(container).toMatchSnapshot();
  });

  it('should render correctly for read-only', () => {
    const { container } = renderWithContext(<WithinRange {...props} value={initialValue} />, {
      isReadOnly: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('should set the rule value when a value is changed and input field is blurred', async () => {
    const { getAllByPlaceholderText } = render(<WithinRange {...props} value={initialValue} />);
    const newValue = [5, 15];

    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[0], { target: { value: newValue[0] } });
    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));

    fireEvent.blur(numberInputs[0]);
    await waitFor(() =>
      expect(props.onValueChange).toHaveBeenCalledWith({
        period: initialValue.period,
        value: [newValue[0], initialValue.value[1]],
      })
    );

    props.onValueChange.mockReset();

    fireEvent.change(numberInputs[1], { target: { value: newValue[1] } });
    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));

    fireEvent.blur(numberInputs[1]);
    await waitFor(() =>
      expect(props.onValueChange).toHaveBeenCalledWith({
        period: initialValue.period,
        value: newValue,
      })
    );
  });

  it('should set the rule value when a period is changed', () => {
    const { getByRole, getByText } = render(<WithinRange {...props} value={initialValue} />);
    const newPeriod = PERIODS[1];

    // open the period dropdown:
    const periodDropdown = getByRole('listbox');
    const periodDropdownButton = within(periodDropdown).getByRole('button');
    userEvent.click(periodDropdownButton);

    // select a new period:
    const newPeriodItem = getByText(newPeriod.label);
    userEvent.click(newPeriodItem);

    expect(props.onValueChange).toHaveBeenCalledWith({
      period: newPeriod.value,
      value: initialValue.value,
    });
  });

  it('should display an error when the first value entered is greater than the second', async () => {
    const { getAllByPlaceholderText, findByText } = render(
      <WithinRange {...props} value={initialValue} />
    );
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[0], { target: { value: initialValue.value[1] + 1 } });
    fireEvent.blur(numberInputs[0]);

    const errorMessage = await findByText('The first value must be smaller than the second.');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should display an error when the second value entered is smaller than the first', async () => {
    const { getAllByPlaceholderText, findByText } = render(
      <WithinRange {...props} value={initialValue} />
    );
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.change(numberInputs[1], { target: { value: initialValue.value[0] - 1 } });
    fireEvent.blur(numberInputs[1]);

    const errorMessage = await findByText('The second value must be greater than the first.');

    expect(errorMessage).toBeInTheDocument();
  });

  it('should not update the rule if the number values were not changed', async () => {
    const { getAllByPlaceholderText } = render(<WithinRange {...props} value={initialValue} />);
    const numberInputs = getAllByPlaceholderText('Enter value');

    fireEvent.blur(numberInputs[0]);
    fireEvent.blur(numberInputs[1]);

    await waitFor(() => expect(props.onValueChange).toHaveBeenCalledTimes(0));
  });
});
