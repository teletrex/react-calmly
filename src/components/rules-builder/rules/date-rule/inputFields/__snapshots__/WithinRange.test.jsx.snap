// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WithinRange should render correctly 1`] = `
<div>
  <div
    class="cds--input-field-container"
    data-testid="date-rule-range-of-last"
  >
    <div
      class="cds--number-input-wrapper"
    >
      <div
        class="cds--form-item"
      >
        <div
          class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
        >
          <div
            class="cds--number__input-wrapper"
          >
            <input
              aria-label="Numeric input field with increment and decrement buttons"
              id="TestDateRuleElement-1"
              max="Infinity"
              min="0"
              name="TestDateRuleElement-1"
              pattern="[0-9]*"
              placeholder="Enter value"
              step="1"
              type="number"
              value=""
            />
            <div
              class="cds--number__controls"
            >
              <button
                aria-atomic="true"
                aria-label="Increment number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn up-icon"
                focusable="false"
                tabindex="-1"
                title="Increment number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="up-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 4l4-4 4 4z"
                  />
                </svg>
              </button>
              <button
                aria-atomic="true"
                aria-label="Decrement number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn down-icon"
                focusable="false"
                tabindex="-1"
                title="Decrement number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="down-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 0L4 4 0 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>

    </div>
    <span
      class="cds--text-space-left cds--text-space-right cds--rule-text"
    >
      to
    </span>
    <div
      class="cds--number-input-wrapper"
    >
      <div
        class="cds--form-item"
      >
        <div
          class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
        >
          <div
            class="cds--number__input-wrapper"
          >
            <input
              aria-label="Numeric input field with increment and decrement buttons"
              id="TestDateRuleElement-2"
              max="Infinity"
              min="0"
              name="TestDateRuleElement-2"
              pattern="[0-9]*"
              placeholder="Enter value"
              step="1"
              type="number"
              value=""
            />
            <div
              class="cds--number__controls"
            >
              <button
                aria-atomic="true"
                aria-label="Increment number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn up-icon"
                focusable="false"
                tabindex="-1"
                title="Increment number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="up-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 4l4-4 4 4z"
                  />
                </svg>
              </button>
              <button
                aria-atomic="true"
                aria-label="Decrement number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn down-icon"
                focusable="false"
                tabindex="-1"
                title="Decrement number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="down-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 0L4 4 0 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>

    </div>
    <div
      class="bx-proposal-dropdown-wrapper"
    >
      <div
        class="bx-proposal-dropdown"
        role="listbox"
      >
        <span>
          <div
            aria-label="Select period"
            class="bx-proposal-dropdown__field"
            role="button"
            tabindex="0"
          >
            <span
              class="bx-proposal-dropdown__field__label"
            >
              days
            </span>
            <div
              class="bx-proposal-dropdown__field__menu-icon"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                />
              </svg>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`WithinRange should render correctly for read-only 1`] = `
<div>
  <span
    class="cds--rule-value-readonly"
  >
    <span
      class="cds--text-space-left"
    >
      2
    </span>
    <span
      class="cds--text-space-left"
    >
      to
    </span>
    <span
      class="cds--text-space-left"
    >
      7
    </span>
    <span
      class="cds--text-space-left"
    >
      months
    </span>
  </span>
</div>
`;

exports[`WithinRange should render correctly with a provided value 1`] = `
<div>
  <div
    class="cds--input-field-container"
    data-testid="date-rule-range-of-last"
  >
    <div
      class="cds--number-input-wrapper"
    >
      <div
        class="cds--form-item"
      >
        <div
          class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
        >
          <div
            class="cds--number__input-wrapper"
          >
            <input
              aria-label="Numeric input field with increment and decrement buttons"
              id="TestDateRuleElement-1"
              max="Infinity"
              min="0"
              name="TestDateRuleElement-1"
              pattern="[0-9]*"
              placeholder="Enter value"
              step="1"
              type="number"
              value="2"
            />
            <div
              class="cds--number__controls"
            >
              <button
                aria-atomic="true"
                aria-label="Increment number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn up-icon"
                focusable="false"
                tabindex="-1"
                title="Increment number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="up-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 4l4-4 4 4z"
                  />
                </svg>
              </button>
              <button
                aria-atomic="true"
                aria-label="Decrement number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn down-icon"
                focusable="false"
                tabindex="-1"
                title="Decrement number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="down-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 0L4 4 0 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>

    </div>
    <span
      class="cds--text-space-left cds--text-space-right cds--rule-text"
    >
      to
    </span>
    <div
      class="cds--number-input-wrapper"
    >
      <div
        class="cds--form-item"
      >
        <div
          class="cds--number cds--number--helpertext cds--number-input-wrapper_carbon-input"
        >
          <div
            class="cds--number__input-wrapper"
          >
            <input
              aria-label="Numeric input field with increment and decrement buttons"
              id="TestDateRuleElement-2"
              max="Infinity"
              min="0"
              name="TestDateRuleElement-2"
              pattern="[0-9]*"
              placeholder="Enter value"
              step="1"
              type="number"
              value="7"
            />
            <div
              class="cds--number__controls"
            >
              <button
                aria-atomic="true"
                aria-label="Increment number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn up-icon"
                focusable="false"
                tabindex="-1"
                title="Increment number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="up-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M0 4l4-4 4 4z"
                  />
                </svg>
              </button>
              <button
                aria-atomic="true"
                aria-label="Decrement number. Press Space or Enter in order to use icon."
                aria-live="polite"
                class="cds--number__control-btn down-icon"
                focusable="false"
                tabindex="-1"
                title="Decrement number. Press Space or Enter in order to use icon."
                type="button"
                unselectable="on"
              >
                <svg
                  aria-hidden="true"
                  class="down-icon"
                  focusable="false"
                  height="4"
                  preserveAspectRatio="xMidYMid meet"
                  style="will-change: transform;"
                  viewBox="0 0 8 4"
                  width="8"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M8 0L4 4 0 0z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
      </div>

    </div>
    <div
      class="bx-proposal-dropdown-wrapper"
    >
      <div
        class="bx-proposal-dropdown"
        role="listbox"
      >
        <span>
          <div
            aria-label="Select period"
            class="bx-proposal-dropdown__field"
            role="button"
            tabindex="0"
          >
            <span
              class="bx-proposal-dropdown__field__label"
            >
              months
            </span>
            <div
              class="bx-proposal-dropdown__field__menu-icon"
            >
              <svg
                aria-hidden="true"
                focusable="false"
                height="16"
                preserveAspectRatio="xMidYMid meet"
                style="will-change: transform;"
                viewBox="0 0 16 16"
                width="16"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 11L3 6l.7-.7L8 9.6l4.3-4.3.7.7z"
                />
              </svg>
            </div>
          </div>
        </span>
      </div>
    </div>
  </div>
</div>
`;
